### ISSUE TITLE
OMG this is nearly perfect but....

### ISSUE THREAD

**Issue by tacticalbacon89** [Yersi88/WMR-and-Vive-Tracker#2]
@Yersi88 its the jitters man it is literally the only thing stopping this from being the absolute perfect solution!!!!!!!
1.  the tracking is great even with the 3.0 vive tracker on the dongle (i haven't tried with usb cable yet)
  2. the offset customizability can be dialed in to almost perfection  using emulator with patch (and they don't have play space drift!!!!!!!)
  3. the index controllers  feel like they are where they belong (once dialed in) and behave as expected
  
 the only thing wrong is the jittery view when moving your head quickly 
if someone could add motion smoothing or camera smoothing to this solution it would be the best solution possible

**Comment by Yersi88** [comment]
Hey! Thanks for the feedback, most appreciated! 

The thing with jitter is that not everyone is getting it. I never noticed any  jitter so I can't really begin to solve problem that I can't see. It appears to be system dependent issue. Perhaps you can include details about your system?

**Comment by tacticalbacon89** [comment]
@Yersi88 
Asrock x99 Tiachi - motherboard
32gb ddr4 - ram
Asus strix RTX 3090 OC -GPU
5820k Haswell-E  - CPU - AIO water cooled- overclocked to 5.0 stable
1000W PSU
dual monitor setup + VR headset
Windows 10- 64bit 

Reverb G2 
valve index controllers 
1- 3.0 vive tracker connected via Quest 2 link cable (this has a huge positive effect on jitters but doesn't completely eliminate it )
2 -   2.0  base stations
2 - tundra vive dongles
your remixed hmd mount

SteamVR ingame visual setting - are at default with resolution set to Auto  
the only setting i changed was SteamVR Home as i dont like it, i prefer the void lol

GAMES PLAYED where Jitters are experienced  in any magnitude these are also the only game i have played so far
BoneWorks
Beatsaber
Half-life Alyx
Pavlov VR
Phasmapobia  

they all seem to produce the same magnitudes of jitters i dont think Games or FPS are a factor maybe alittle but not that i can tell honestly

it feels like this method needs some form of camera movement smoothing or motion smoothing algorithm or atleast as a option for those who do get the jitters. 
but literally everything else about this method works flawlessly (surprisingly way better than i expected)

if someone could create a separate tool to smooth the motions of the camera it would make this method the absolute best solution for the G2 controller alternatives
this method is nearly perfect almost feels native the jitters are the only thing holding it back 

1. the fishbowl/warping effect and camera being off is easily fixed with Input Emulator offsets 
2. grey screen flashes easily fixed with more base stations/better station setup 
3.  room scale setup is easily fixed with OVRAS and saving profiles
4. now we just need smoothing tool for jitters

the jitter  happens mostly when moving your head at a fast rate and if using dongle with tracker at ANY rate of speed 
(i can walk around at medium pace with head straight and the jitters are not bad at all definitely 100% playable like this)
its when you twist your head mostly on roll, pitch, and yaw axis's it seems lateral movement doesn't produce the jitter quite as pronounced
you experience this mostly when looking at game UI's or SteamVR UI but also it happens in gameplay aswell your just distracted and don't see it as much unless it gets bad 

if i could visually describe the jitters
 it looks like  as if  the frames are trying to play catch up in micro seconds,  almost as if you have slightly  low FPS or have your MSAA too high, its not effecting the world itself but rather your view of the world 
my games are not lagging from my perspective
 if i stand still or not twist my head fast everything is butter smooth  and looks like a dream come true 
controllers do not have this effect independently, i can swing them around they don't jitter, other than normal light house tech problems,   it seems to be remote to virtual camera/tracker combo

**Comment by Yersi88** [comment]
Thanks for the detailed experience report here! 

One reason there might be the jitter is that vive trackers run at 120Hz refresh rate and native SteamVR tracked headsets run at over 10x that! 

I'm not sure how I'd go about fixing the jitter honestly. I'm not sure if simple filter (Kanban?) would do the trick and even if it did idk how to implement it let alone how to implement it in such manner that it wouldn't add any latency, I'm afraid that any processing like that will add latency no matter what we do.  

Also it's kind of hard for me to address the jitters because I don't experience them at all!

**Comment by tacticalbacon89** [comment]
im afraid to buy another vive tracker as the cost of this setup is increasing more and more  but i want to test and see if using a 2.0 instead of a 3.0 would solve this in my particular case.

im not sure if it is a refresh rate issue as what ever refresh rate your device is capable of is what you will experience and 120hz is a decently smooth refresh rate by monitor standards also being as the jitter doesn't occur in all instances of movement it seems  to be localized to the pitch roll and yaw axis's at least in its most drastic form (is there a way to bump a vive trackers refresh rate? or can that be achieved for a programing standpoint  ) ,  if so maybe a brute force of the sort and or force the virtual camera to handshake comfortably at 120Hz 
 im not sure exactly how that hand shake works between virtual camera and tracked device, i assume its just raw distance data from the tracker to steamvr then steamvr calculates and refreshes its position but maybe the secret to this issue is in the parameters of the virtual camera and not the tracker

**Comment by tacticalbacon89** [comment]
a filter of some kind to turn IMU data into smoothed velocities and angular velocities for prediction

This was qouted to my by another software developer and would be a solution @Yersi88

**Comment by tacticalbacon89** [comment]
Would this help in any way https://github.com/balzer82/Kalman

**Comment by Yersi88** [comment]
> Would this help in any way https://github.com/balzer82/Kalman

No, I know what Kalman is. But to implement it you need openvr driver that will take data from tracker and applies Kalman than creates new virtual device with smoothed output. And that might not even work good...



> a filter of some kind to turn IMU data into smoothed velocities and angular velocities for prediction
> 
> This was qouted to my by another software developer and would be a solution @Yersi88

I'm afraid that this might not work, Kalman is not prediction, it works with data 'from past' and it would introduce lag. I'm sure openvr has prediction models but you can't put Kalman in front of that (easily). That would require rewriting whole firmware of vive tracker which is not even open source as far as I know. Rewriting firmware and editing raw data from imu, that never gonna happen, we might have more luck bodging our own tracking hardware together....


Anyway, As I said before this is as good as it's gonna get for now, unless there's a C++ / OVR dev out there that wats to collaborate with me on this.
